### PR TITLE
Port basic non-security tests to k8s

### PR DESCRIPTION
**Pull request by ericgribkoff**
Adds baseline K8S tests equivalent to `round_robin`, `change_backend_service`, `remove_instance_group`, `gentle_failover`, `secondary_locality_gets_no_requests_on_partial_primary_failure`, and `secondary_locality_gets_requests_on_primary_failure` from the existing GCE test suite. These, along with `backends_restart` (which doesn't seem to have a clear analog in a K8S deployment), constitute the "original" suite of GCE tests and serve as a good starting point for the requisite infrastructure additions to the K8S test runner.

These infra additions include:

1) Support for deployments in another region (=> a second cluster/kube context)
1) Support for additional backend services
1) Support for additional NEGs
1) Support for replica_count > 1
1) Support for patching URL maps
1) `maxRatePerEndpoint` is configurable per-backend service (required for failover tests to function)
1) `assertRpcsEventuallyGoToGivenServers` method wrapped by tenacity for verifying expected RPC distribution

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/26360)

### CODE REVIEW

**Review by lidizheng — COMMENTED** [review]
Great work! I think the only overlapping part for this PR and https://github.com/grpc/grpc/pull/26367 is around backend services. I can use this PR's API after it is merged, since this API is more powerful.

Nothing major in the comments. Can we have any sort of test run on Kokoro?

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26360#-:-Mb867VN0ypTiIJX4S1j:b-stus5a)* status: 0 of 13 files reviewed, 6 unresolved discussions (waiting on @ericgribkoff and @sergiitk)

___
*[tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py, line 20 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb8-oOl9Fzjk4AqaFMj:-Mb8-oOl9Fzjk4AqaFMk:b-ybpxv8) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py#L20)):*
> ```Python
> import time
> from typing import List
> from typing import Optional, Tuple
> ```

Merge the `typing` imports?
___
*[tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py, line 304 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb80AJNCve3-rIXlyEW:-Mb80AJNCve3-rIXlyEX:b-cbuv0w) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py#L304)):*
> ```Python
>             reuse_namespace=self.server_namespace == self.client_namespace)
> 
>     def startTestServer(self, server_runner, replica_count=1, **kwargs) -> List[
> ```

nit: Many new methods in this file has missing type annotations. This line:`server_runner: KubernetesServerRunner, replica_count: int=1`
___
*[tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director.py, line 153 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb80pVHBxb-CWFGeXYH:-Mb80pVHBxb-CWFGeXYI:bdr9ziv) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director.py#L153)):*
> ```Python
> 
>     def create_backend_service(
>         self, name: str = BACKEND_SERVICE_NAME,
> ```

We may also want alternative constants for backend service names.
___
*[tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director.py, line 162 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb81St5ERO_dvxsAb78:-Mb81St5ERO_dvxsAb79:b-wlzp7q) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director.py#L162)):*
> ```Python
>         resource = self.compute.create_backend_service_traffic_director(
>             ns_name, health_check=self.health_check, protocol=protocol)
>         self.backend_services[name] = resource
> ```

nit: A backend service dictionary, but the name doesn't match the actual resource name. Maybe we should add a `get_backend_service` method?
___
*[tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director.py, line 259 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb8282M5R6upkQ40HdW:-Mb8282M5R6upkQ40HdX:ba5kaxb) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/infrastructure/traffic_director.py#L259)):*
> ```Python
>         name = self._ns_name(self.URL_MAP_NAME)
>         matcher_name = self._ns_name(self.URL_MAP_PATH_MATCHER_NAME)
>         logger.info('Creating URL map "%s": %s -> %s', name, src_address,
> ```

Patching.
___
*[tools/run_tests/xds_k8s_test_driver/tests/change_backend_service_test.py, line 1 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb834Sp59AVmQJrHvLA:-Mb834Sp59AVmQJrHvLB:bw023hr) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/tests/change_backend_service_test.py#L1)):*
> ```Python
> # Copyright 2020 gRPC authors.
> ```

s/2020/2021
Same for other test files

**Review by sergiitk — COMMENTED** [review]
To make our life a bit simpler, and skip avoidable style comments, I'll add some helper tools to run yapf and isort. Hold on.

> Reviewed 2 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26360#-:-MczcCnh0uX33aXXHQHp:b36hs2u)* status: 2 of 13 files reviewed, 7 unresolved discussions (waiting on @ericgribkoff, @lidizheng, and @sergiitk)

___
*[tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_flags.py, line 22 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mcz_YC33HApbBbbWPbY:-Mcz_YC33HApbBbbWPbZ:bk668y5) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_flags.py#L22)):*
> ```Python
> SECONDARY_KUBE_CONTEXT = flags.DEFINE_string("secondary_kube_context",
>                                    default=None,
>                                    help="Secondary kubectl context to use for cluster in another region")
> ```

This definitely will get reformatted by yapf. I'll add a helper to make things simpler for us.
___
*[tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py, line 20 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Mb8-oOl9Fzjk4AqaFMj:-Mczbgm4BXNT4D-IX3M2:bc7vaw7) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py#L20)):*
<details><summary><i>Previously, lidizheng (Lidi Zheng) wrote…</i></summary><blockquote>

Merge the `typing` imports?
</blockquote></details>

This should be simpler for us. See my comment on `isort` here: https://github.com/grpc/grpc/pull/26367#pullrequestreview-691594043 (don't know how to link it directly).

**Review by sergiitk — COMMENTED** [review]
This looks pretty good. A lot of nice additions and improvements.  

As long as this doesn't break security tests, and is reconciled with @lidizheng's changes, this is LGTM. No serious design-level or conceptual issues. This is sold.

If I'm unavailable for follow up reviews, I have my full trust in @lidizheng to provide you with the feedback, and approve this when it's ready.

> Reviewed 7 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26360#-:-Md-NC_s3rRlEZqnnMKi:b8v9ktw)* status: 9 of 13 files reviewed, 9 unresolved discussions (waiting on @ericgribkoff, @lidizheng, and @sergiitk)

___
*[tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py, line 119 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Md-IzA46PNCUTz9_ZlJ:-Md-IzA46PNCUTz9_ZlK:bvnxmao) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py#L119)):*
> ```
> self.server_runners
> ```

This probably should be a class, something like

```py
@dataclass
class ServerRuners:
  primary: KubernetesServerRunner
  secondary: KubernetesServerRunner
```

Not necessary a dataclass, just to illustrate the point
___
*[tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py, line 188 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Md-JaBd8hofmh7zZ0iS:-Md-JaBd8hofmh7zZ0iT:b8q0eh2) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py#L188)):*
> ```Python
>             msg=f'Expected all RPCs to succeed: {failed} of {num_rpcs} failed')
> 
>     @tenacity.retry(stop=tenacity.stop_after_delay(TD_CONFIG_MAX_WAIT_SEC),
> ```

This is ok, but take a look at `framework.helpers.retryers.constant_retryer` - this makes it easier to parametrize the wait times.
___
*[tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py, line 62 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Md-KZxxD6ZbF4oADstv:-Md-KZxxD6ZbF4oADstw:b-79n8yv) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py#L62)):*
> ```Python
>         self.server_id = server_id
>         self.xds_host, self.xds_port = xds_host, xds_port
>         self.pod_name = pod_name
> ```

`XdsTestServer` is supposed to represent [xds test server](https://github.com/grpc/grpc/blob/master/doc/xds-test-descriptions.md#server) available via remote connection. It shouldn't be coupled to a specific platform running it. `pod.metadata.name` is specific to k8s.
However, I'm not opposed to adding something like this. Maybe hostname?

I know it's totally unclear from the design (my fault for mixing `XdsTestServer` with kubernetes runners). It's not the first time this is brought up: https://github.com/grpc/grpc/pull/26367#discussion_r640848496
___
*[tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py, line 298 at r1](https://reviewable.io/reviews/grpc/grpc/26360#-Md-MvdDB0M6MAHbHtn_:-Md-MvdDB0M6MAHbHtna:b-9dim3p) ([raw file](https://github.com/grpc/grpc/blob/9bf860b1029ae8ee77d7be7895a0078545eda191/tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py#L298)):*
> ```Python
>                     'LOCAL DEV MODE: Enabling port forwarding to %s:%s using local port %s',
>                     pod_ip, maintenance_port, local_port)
>                 self.port_forwarders.append(self.k8s_namespace.port_forward_pod(
> ```

Wow, this is great. Thanks for solving this.

**Review by sergiitk — COMMENTED** [review]
@ericgribkoff added helper tools to run yapf and isort. They basically do what I do manually. @lidizheng will look up into automating them. Until then, feel free to use, should be helpful with passing that Sanity Check job.

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/26360#-:-Md-fO1a4mws3qvIhppf:b-qs18be)* status: 9 of 13 files reviewed, 9 unresolved discussions (waiting on @ericgribkoff and @lidizheng)

**Comment by ericgribkoff** [comment]
Master has changed quite a bit from when this branch was originally cut, particularly framework/infrastructure/traffic_director.py, which now has `alternative_backend_service` and `affinity_backend_service` so I'm removing the part of the change here that switched the original `backend_service` into a map of `backend_services`. This changes the signature of a number of methods in the new tests and elsewhere, so going through those now to try to resolve the conflicts (the merge commit got rid of a bunch, but some of the new code here is calling functions with the wrong signature now). Will update this again once everything is resolved.

**Comment by sergiitk** [comment]
Yes, there was even also a couple of features backported from your PR  :) 
https://github.com/grpc/grpc/pull/26759

Thanks for the contribution, and apologies for the merge/rebase getting complicated.

**Comment by ericgribkoff** [comment]
My attempt to make the history still followable failed miserably. Basically everything is in https://github.com/grpc/grpc/pull/26360/commits/7ee7f506f95d9fd62b3908fbc026a6e003c37282, redone on top of the current master, but I was unable to trick git into showing the diff between this and the last-reviewed commit of this PR, so everything shows again as entirely new changes despite being mostly just a small modification of what was already reviewed to bring it inline with the current master.

If there's some other way I can structure the merge with master to make it easier to (re)review this, let me know. Everything passes locally and I just kicked off a run on Kokoro (sans the failover test, as I'll need to add another k8s cluster in a second region for this test case). Will update with the run results once finished.

File: tools/run_tests/xds_k8s_test_driver/tests/failover_test.py
Comment (ericgribkoff) [inline tools/run_tests/xds_k8s_test_driver/tests/failover_test.py:41]: @sergiitk So this proved fraught. Originally I had the suffix here as `-alternate-region`, which didn't show any problems in any of the operations of the framework (as far as I could determine from the logs, but I could easily have missed something) but the deployment in this runner was never moving to ready. Eventually I was able to discover:

```
$ kubectl get svcneg -o yaml
     message: 'googleapi: Error 400: Invalid value for field ''networkEndpointGroup'':
        ''xds-k8s-security-server-20210729-0938-k5ad1-psm-grpc-server-alternate-region''.
        Must be a match of regex ''(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)|[1-9][0-9]{0,19}'',
```

But this took me (admittedly very new to k8s) quite a while of digging around to even know to check this. I think I eventually noticed that there was no NEG for the deployment, and that led to the above kubectl check that revealed the error. But I couldn't identify a particular operation of the framework that actually failed, so it wasn't clear to me if/where we could put a check to help avoid future-me from running into this same too-long-of-a-name problem again and not knowing what the underlying problem is. Curious if you might have a better idea of what step actually led to this and were it could be checked for.

Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/tests/failover_test.py:41]: Ooof, I see how this could've been frustrating. 

1. One hard-to-predict thing that changed with adding the "multiple framework runs at the same time" feature (#26542) is added resource suffix, with date-time and nonce, f.e. `-20210729-0938-k5ad1`. This is needed to make each run unique. However this added 20 chars to each resource name. And different resource types have different char limits. You can see in the error you're getting, there's a 61 char limit on NEG name: `{0,61}`. In many cases these char limits are implicit, and the only way to know they exist, is to just run the script, and check for errors related to the name field. That's the reason I tried to cut down on char limit, and replaced `-alternate` with `-alt` in all places I found.
2. The problem with debugging NEGs is that they are a [custom k8s extension](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) made by GCP for GCP. There's a bunch on opaque machinery invisible to k8s. That's also why it takes forever to delete a k8s namespace, when there are live GCP resources using a NEG associated with a service in the namespace (f.e. this happens when cleanup is done in incorrect order)
3. My recent life learning is that most cases I get stuck in debugging, it could've spared myself some time by asking someone who worked on that thing I'm debugging. I'm still pretty bad at applying this lesson though 😄

**Comment by ericgribkoff** [comment]
change_backend_service, round_robin, and remove_neg tests all passed on an ad-hoc Kokoro run: https://fusion2.corp.google.com/invocations/3eb3df36-d30b-472d-b3fd-21d7726a2191

Will add the secondary cluster and rerun failover_test tomorrow (already verified on my own gke project)

**Comment by ericgribkoff** [comment]
The new tests all passed on Kokoro: https://source.cloud.google.com/results/invocations/e6a231fa-f4bd-4076-ab9f-b2f760007d19/targets

The job itself hit the 120 minute timeout however. @sergiitk Should I just increase the timeout, or would you rather structure the jobs differently so we're not running baseline -> new tests -> security all in the same job?

**Comment by sergiitk** [comment]
@ericgribkoff I think we should run it as a separate job, similar to URL map: https://github.com/grpc/grpc/blob/master/tools/internal_ci/linux/grpc_xds_url_map.cfg
 
Or as a part of URL map tests. CC @lidizheng — what do you think?

But definitely not with PSM Security, because ideally we want to run them against separate GKE cluster.

**Comment by ericgribkoff** [comment]
I'll set up a separate job, unless there's some reason to keep these bundled with url map (since the url map tests do some configuration/scheduling work to share resources across runs/languages, and this doesn't really apply to these tests)

**Comment by ericgribkoff** [comment]
Rolled back the job config file changes, will add the job for these tests in a separate PR. This is ready to rereview and move forward.

**Review by sergiitk — APPROVED** [review]
LGTM!

**Pull request merged**
